### PR TITLE
Heatmap

### PR DESCRIPTION
**Pull request by Rygone**
Add the Heatmap
Fix EditProfileActivityTest
Update Observable with `map` and `update` method

### CODE REVIEW

File: app/src/main/java/com/github/sdpteam15/polyevents/fragments/MapsFragment.kt [hunk 1]
```diff
@@ -122,11 +125,31 @@ class MapsFragment : Fragment(), OnMapReadyCallback, OnPolylineClickListener,
...
+                val task = object : TimerTask() {
+                    override fun run() {
+                        val locations = ObservableList<LatLng>()
+                        Database.currentDatabase.heatmapDatabase!!.getLocations(locations)
+                        locations.observeOnce {
+                            GoogleMapHelper.addHeatMap(it.value)
+                        }
+                    }
+                }
+                timerHeatmap?.schedule(task, 0, 15 * 1000)
```
Comment (Mathematou) [inline app/src/main/java/com/github/sdpteam15/polyevents/fragments/MapsFragment.kt:144]: We could have these magic numbers as constants

File: app/src/main/java/com/github/sdpteam15/polyevents/service/Timer_Service.kt [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+    override fun onBind(intent: Intent): IBinder? {
+        return null
+    }
+
+    override fun onCreate() {
+        super.onCreate()
+        instance.value = this
+
+        mTimer = Timer()
+        mTimer!!.scheduleAtFixedRate(TimeDisplayTimerTask(), 0 , 30 * 1000)
```
Comment (Mathematou) [inline app/src/main/java/com/github/sdpteam15/polyevents/service/Timer_Service.kt:30]: Same here, magic numbers could be constants

**Review by Mathematou — APPROVED** [review]
Nice, tell me when your tests are done :)

**Review by Mathematou — COMMENTED** [review]
Nice

File: app/src/main/java/com/github/sdpteam15/polyevents/MainActivity.kt [hunk 3]
```diff
@@ -46,6 +48,20 @@ class MainActivity : AppCompatActivity() {
         super.onCreate(savedInstanceState)
         setContentView(R.layout.activity_main)
 
+        Settings.mainActivity = this
+        Settings.isLoaded = Settings.isLoaded
```
Comment (Feez2403) [inline app/src/main/java/com/github/sdpteam15/polyevents/MainActivity.kt:52]: I think you can delete this line

File: app/src/main/java/com/github/sdpteam15/polyevents/database/observe/Observable.kt [hunk 4]
```diff
@@ -1,24 +1,40 @@
...
+import androidx.lifecycle.LifecycleObserver
 import androidx.lifecycle.LifecycleOwner
-import com.github.sdpteam15.polyevents.helper.HelperFunctions.observeOnDestroy
+import androidx.lifecycle.OnLifecycleEvent
 import com.github.sdpteam15.polyevents.helper.HelperFunctions.run
 
 /**
  * Observable live data of type T
  * @param
  */
```
Comment (Feez2403) [inline app/src/main/java/com/github/sdpteam15/polyevents/database/observe/Observable.kt:12]:
```suggestion
/**
 * Data that notify a set of observers on a modification
 * When adding observers and passing a lifecycleOwner in the parameters, the added observer will not be notified if the lifecycle is destroyed.
 * This is useful for example when an activity is stopped, as we don't need to update data on this closed activity.
 * If we don't give a Lifecycle in parameters, the observer will keep beeing notified even if the activity is stopped.
 * To unsubscribe from the Observable, we take take the reference of the corresponding "remove" function
 * We get the "remove" function by taking the return value of the observe method, for example :
 *
 *      val ret : ThenOrRemove = observable.observe { ... }
 *      ...
 *      do something
 *      ...
 *      //on a special event stop observing
 *      ret.remove()
 *
 *  The chain operator "then" from this return function can be used to add multiple observers in a convenient way
 *  For example in an activity:
 *
 *      observable.map(this){ ... }.then.observe(this){ ... }.then.map(this){}.then.observeOnce(this){ ... }
 *
 * When setting a value, we use an "UpdateValue" object which is a pair containing the new value,
 * and eventually a reference to the sender object, which can be used when we need to know who sent the update.
 *
 * @param value initial value of the data
 * @param sender last object that modified the data
 */
```

File: app/src/main/java/com/github/sdpteam15/polyevents/database/observe/Observable.kt [hunk 5]
```diff
@@ -1,24 +1,40 @@
...
 class Observable<T>(value: T? = null, sender: Any? = null) {
-    constructor(value: T? = null) : this(value, null)
-    constructor() : this(null)
+    class AfterRemovable<U>(val then: U, val remove: () -> Boolean)
+    class UpdateValue<T>(val value: T, val sender: Any?)
```
Comment (Feez2403) [inline app/src/main/java/com/github/sdpteam15/polyevents/database/observe/Observable.kt:15]: /**
 * Data that notify a set of observers on a modification
 * When adding observers and passing a lifecycleOwner in the parameters, the added observer will not be notified if the lifecycle is destroyed.
 * This is useful for example when an activity is stopped, as we don't need to update data on this closed activity.
 * If we don't give a Lifecycle in parameters, the observer will keep beeing notified even if the activity is stopped.
 * To unsubscribe from the Observable, we take take the reference of the corresponding "remove" function
 * We get the "remove" function by taking the return value of the observe method, for example :
 *
 *      val ret : ThenOrRemove = observable.observe { ... }
 *      ...
 *      do something
 *      ...
 *      //on a special event stop observing
 *      ret.remove()
 *
 *  The chain operator "then" from this return function can be used to add multiple observers in a convenient way
 *  For example in an activity:
 *
 *      observable.map(this){ ... }.then.observe(this){ ... }.then.map(this){}.then.observeOnce(this){ ... }
 *
 * When setting a value, we use an "UpdateValue" object which is a pair containing the new value,
 * and eventually a reference to the sender object, which can be used when we need to know who sent the update.
 *
 * @param value initial value of the data
 * @param sender last object that modified the data
 */
class Observable<T>(value: T? = null, sender: Any? = null) {

    /**
     * The return value of each function of an observable
     *
     * @param then returns a reference of the observable after calling a transformation
     * @param remove function to remove the observer from the list
     */
    class AfterRemovable<U>(val then: U, val remove: () -> Boolean)

    /**
     * We use an UpdateValue object each time we want to set a new value for the data.
     * It contains the new value and eventually the object that set the new value.
     *
     * @param value new value of the data
     * @param sender function to remove the observer from the list
     */
    class UpdateValue<T>(val value: T, val sender: Any?)

File: app/src/main/java/com/github/sdpteam15/polyevents/database/observe/Observable.kt [hunk 6]
```diff
@@ -1,24 +1,40 @@
...
 import androidx.lifecycle.LifecycleOwner
-import com.github.sdpteam15.polyevents.helper.HelperFunctions.observeOnDestroy
+import androidx.lifecycle.OnLifecycleEvent
 import com.github.sdpteam15.polyevents.helper.HelperFunctions.run
 
 /**
  * Observable live data of type T
  * @param
  */
 class Observable<T>(value: T? = null, sender: Any? = null) {
```
Comment (Feez2403) [inline app/src/main/java/com/github/sdpteam15/polyevents/database/observe/Observable.kt:13]:
```suggestion
/**
 * Data that notify a set of observers on a modification
 * When adding observers and passing a lifecycleOwner in the parameters, the added observer will not be notified if the lifecycle is destroyed.
 * This is useful for example when an activity is stopped, as we don't need to update data on this closed activity.
 * If we don't give a Lifecycle in parameters, the observer will keep beeing notified even if the activity is stopped.
 * To unsubscribe from the Observable, we take take the reference of the corresponding "remove" function
 * We get the "remove" function by taking the return value of the observe method, for example :
 *
 *      val ret : ThenOrRemove = observable.observe { ... }
 *      ...
 *      do something
 *      ...
 *      //on a special event stop observing
 *      ret.remove()
 *
 *  The chain operator "then" from this return function can be used to add multiple observers in a convenient way
 *  For example in an activity:
 *
 *      observable.map(this){ ... }.then.observe(this){ ... }.then.map(this){}.then.observeOnce(this){ ... }
 *
 * When setting a value, we use an "UpdateValue" object which is a pair containing the new value,
 * and eventually a reference to the sender object, which can be used when we need to know who sent the update.
 *
 * @param value initial value of the data
 * @param sender last object that modified the data
 */
class Observable<T>(value: T? = null, sender: Any? = null) {
```

File: app/src/main/java/com/github/sdpteam15/polyevents/database/observe/Observable.kt (hunk 5 above)
Comment (Feez2403) [inline app/src/main/java/com/github/sdpteam15/polyevents/database/observe/Observable.kt:15]:
```suggestion
/**
     * The return value of each function of an observable
     *
     * @param then returns a reference of the observable after calling a transformation
     * @param remove function to remove the observer from the list
     */
    class AfterRemovable<U>(val then: U, val remove: () -> Boolean)

    /**
     * We use an UpdateValue object each time we want to set a new value for the data.
     * It contains the new value and eventually the object that set the new value.
     *
     * @param value new value of the data
     * @param sender function to remove the observer from the list
     */
    class UpdateValue<T>(val value: T, val sender: Any?)
```

File: app/src/test/java/com/github/sdpteam15/polyevents/database/HelperTestFunction.kt [hunk 7]
```diff
@@ -0,0 +1,189 @@
...
+
+import com.github.sdpteam15.polyevents.database.observe.Observable
+import com.github.sdpteam15.polyevents.database.observe.ObservableList
+import com.github.sdpteam15.polyevents.util.AdapterFromDocumentInterface
+import com.github.sdpteam15.polyevents.util.AdapterToDocumentInterface
+import org.mockito.Mockito
+import org.mockito.kotlin.anyOrNull
+import java.util.*
+import org.mockito.Mockito.`when` as When
+
```
Comment (Feez2403) [inline app/src/test/java/com/github/sdpteam15/polyevents/database/HelperTestFunction.kt:11]: comment

**Pull request merged**